pkg/operator/controller/credentials-secret/credentials_secret.go

```diff
@@ -95,14 +96,34 @@ func (r *reconciler) createCredentialsSecret(ctx context.Context, secret *corev1
 }
 
 // desiredCredentialsSecret returns the desired destination secret.
-func desiredCredentialsSecret(sourceSecret *corev1.Secret, destName types.NamespacedName) *corev1.Secret {
-	return &corev1.Secret{
+func desiredCredentialsSecret(sourceSecret *corev1.Secret, destName types.NamespacedName, extDNS *operatorv1alpha1.ExternalDNS) *corev1.Secret {
+	var secret *corev1.Secret
+	var data map[string][]byte
+	secret = &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      destName.Name,
 			Namespace: destName.Namespace,
 		},
-		Data: sourceSecret.Data,
 	}
+	data = sourceSecret.Data
```

Comment:
You know that `data` and `sourceSecret.Data` are now pointing to the same dataset?
When you change `data` this changes `sourceData` too.

I think this method needs to be refactored (to avoid the reference problem and the need to `delete` keys):
- `secret.Data` can be added with new values directly without this intermediate `data` variable
- for AWS case you need to do a `for` loop to get all key/values from `sourceSecret` to avoid the same reference problem as here